File: typescript/frontend/src/api/job.ts
import { match } from "ts-pattern";
import { z } from "zod";

export const JobStatus = z.enum([
  "PENDING",
  "RUNNING",
  "CANCELLED",
  "SUCCEEDED",
  "FAILED",
]);

export function isJobCompleted(status: z.infer<typeof JobStatus>): boolean {
  return match(status)
    .with("PENDING", () => false)
    .with("RUNNING", () => false)
    .with("CANCELLED", () => true)
    .with("SUCCEEDED", () => true)
    .with("FAILED", () => true)
    .exhaustive();
}

export const PendingJobInfo = z.object({
  type: z.literal(JobStatus.Values.PENDING),
});

export const RunningJobInfo = z.object({
  type: z.literal(JobStatus.Values.RUNNING),
  progress: z
    .object({
      num_updated: z.number(),
      num_total: z.number(),
    })
    .nullable(),
});

export const CancelledJobInfo = z.object({
  type: z.literal(JobStatus.Values.CANCELLED),
});

export const SucceededJobInfo = z.object({
  type: z.literal(JobStatus.Values.SUCCEEDED),
  completed_at: z.string(),
});

export const FailedJobInfo = z.object({
  type: z.literal(JobStatus.Values.FAILED),
  // error: z.string(),
  completed_at: z.string(),
});

export type JobInfo = z.infer<typeof JobInfo>;
export const JobInfo = z.union([
  PendingJobInfo,
  RunningJobInfo,
  CancelledJobInfo,
  SucceededJobInfo,
  FailedJobInfo,
]);

File: python/backend/src/job.py
from datetime import datetime
from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, Field


class JobStatus(str, Enum):
    Pending = "PENDING"
    Running = "RUNNING"
    Cancelled = "CANCELLED"
    Succeeded = "SUCCEEDED"
    Failed = "FAILED"

    def is_completed(self) -> bool:
        match self:
            case JobStatus.Pending | JobStatus.Running:
                return False

            case JobStatus.Cancelled | JobStatus.Succeeded | JobStatus.Failed:
                return True


class PendingJobInfo(BaseModel):
    type: Literal[JobStatus.Pending]


class RunningJobInfo(BaseModel):
    class Progress(BaseModel):
        num_updated: int
        num_total: int

    type: Literal[JobStatus.Running]
    progress: Progress | None = None


class CancelledJobInfo(BaseModel):
    type: Literal[JobStatus.Cancelled]


class SucceededJobInfo(BaseModel):
    type: Literal[JobStatus.Succeeded]
    completed_at: datetime


class FailedJobInfo(BaseModel):
    type: Literal[JobStatus.Failed]
    error: str
    completed_at: datetime


JobInfo = Annotated[
    PendingJobInfo
    | RunningJobInfo
    | CancelledJobInfo
    | SucceededJobInfo
    | FailedJobInfo,
    Field(..., discriminator="type"),
]
